pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,292 @@
+			// The Endpoints resource for a Service has the same name as the Service.
+			if endpoints, ok := obj.(*corev1.Endpoints); ok {
+				return endpoints.Namespace == namespace && endpoints.Name == serviceName
+			}
+			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				if endpoints, ok := tombstone.Obj.(*corev1.Endpoints); ok {
+					return endpoints.Namespace == namespace && endpoints.Name == serviceName
+				}
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				resolver.queue.Add(key)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				// This should not happen: both objects should be Endpoints in the
+				// update event handler.
+				oldEndpoints, ok := oldObj.(*corev1.Endpoints)
+				if !ok {
+					return
+				}
+				newEndpoints, ok := newObj.(*corev1.Endpoints)
+				if !ok {
+					return
+				}
+				// Ignore changes to metadata.
+				if reflect.DeepEqual(newEndpoints.Subsets, oldEndpoints.Subsets) {
+					return
+				}
+				resolver.queue.Add(key)
+			},
+			DeleteFunc: func(obj interface{}) {
+				resolver.queue.Add(key)
+			},
+		},
+	})
+	return resolver
+}
+
+func (r *EndpointResolver) Run(ctx context.Context) {
+	defer r.queue.ShutDown()
+
+	klog.InfoS("Starting controller", "name", r.name)
+	defer klog.InfoS("Shutting down controller", "name", r.name)
+
+	r.informerFactory.Start(ctx.Done())
+	defer r.informerFactory.Shutdown()
+
+	if !cache.WaitForNamedCacheSync(r.name, ctx.Done(), r.serviceListerSynced, r.endpointsListerSynced) {
+		return
+	}
+
+	// We only start one worker for this controller.
+	go wait.Until(r.runWorker, time.Second, ctx.Done())
+
+	<-ctx.Done()
+}
+
+func (r *EndpointResolver) runWorker() {
+	for r.processNextWorkItem() {
+	}
+}
+
+func (r *EndpointResolver) processNextWorkItem() bool {
+	key, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(key)
+
+	if err := r.resolveEndpoint(); err == nil {
+		r.queue.Forget(key)
+	} else {
+		klog.ErrorS(err, "Failed to resolve Service Endpoint, requeuing", "key", key)
+		r.queue.AddRateLimited(key)
+	}
+
+	return true
+}
+
+func (r *EndpointResolver) resolveEndpoint() error {
+	klog.V(2).InfoS("Resolving Endpoint", "service", klog.KRef(r.namespace, r.serviceName))
+	endpointURL, err := proxy.ResolveEndpoint(r.serviceLister, r.endpointsLister, r.namespace, r.serviceName, r.servicePort)
+	// Typically we will get one of these 2 errors (unavailable or not found).
+	// In this case, it makes sense to reset the Endpoint URL to nil and notify listeners.
+	// There is also no need to retry, as we won't find a suitable Endpoint until the Service or
+	// the Endpoints resource is updated in a way that will cause this function to be called again.
+	if errors.IsServiceUnavailable(err) {
+		klog.ErrorS(err, "Cannot resolve endpoint because Service is unavailable", "service", klog.KRef(r.namespace, r.serviceName))
+		r.updateEndpointIfNeeded(nil)
```

Comment:
@tnqn I wanted to bring this to your attention, as it means that `GetAntreaClient` can return `(nil, non-nil error)` when the Antrea Service is not available (e.g. the antrea-controller Pod is restarted). I believe that with the previous behavior, `GetAntreaClient` was guaranteed to always succeed after the first successful call (that doesn't mean that the Service could be accessed successfully).

Let me know what you think. I think this is the right thing to do, and I don't think that should impact existing consumers of `GetAntreaClient`, but I can also remove the calls to `r.updateEndpointIfNeeded(nil)` if you prefer.

Reply:
The change makes sense to me.